### PR TITLE
Multithreaded asset downloading, should be a lot quicker

### PR DESCRIPTION
**Pull request by modmuss50**

### CODE REVIEW

File: src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java [hunk 1]
```diff
@@ -76,18 +80,18 @@ public static void provide(MinecraftProvider minecraftProvider, Project project)
 			}
 		}
 
-		ProgressLogger progressLogger = ProgressLogger.getProgressFactory(project, MinecraftAssetsProvider.class.getName());
-		progressLogger.start("Downloading assets...", "assets");
+		project.getLogger().lifecycle(":downloading assets...");
+
+		Stack<ProgressLogger> loggers = new Stack<>();
+		ExecutorService executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() / 2);
```
Comment (dblsaiko) [inline src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java]: Pretty sure `Runtime.getRuntime().availableProcessors() / 2` will give you 0 when on a single core processor.

**Comment by modmuss50** [comment]
I have aligned the hashes so it looks a lot better when downloading

File: src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java (hunk 1 above)
Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java]: Sloved it.

**Review by dblsaiko — APPROVED** [review]

**Comment by natanfudge** [comment]
How is a network-bound task sped up by multithreading?

**Comment by modmuss50** [comment]
When downloading 1000s of small files a lot of time is spent establishing the network connection. This allows for the bandwidth to be maxed out, in some cases drastically increasing download speed

File: src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java [hunk 2]
```diff
@@ -76,18 +80,18 @@ public static void provide(MinecraftProvider minecraftProvider, Project project)
 			}
 		}
 
-		ProgressLogger progressLogger = ProgressLogger.getProgressFactory(project, MinecraftAssetsProvider.class.getName());
-		progressLogger.start("Downloading assets...", "assets");
+		project.getLogger().lifecycle(":downloading assets...");
+
+		Stack<ProgressLogger> loggers = new Stack<>();
```
Comment (liach) [inline src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java]: Deque please

**Review by liach — CHANGES_REQUESTED** [review]

File: src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java (hunk 2 above)
Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java]: Why, whats the issue with stack?

Comment (Earthcomputer) [reply src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java]: Stack is deprecated in newer versions of Java and will be removed eventually. Deque is the preferred replacement.

Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java]: ah, didnt know that. I will sort that out later. Thanks

Comment (Chocohead) [reply src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java]: `Stack` (and all its other pre-collections friends) are obsolete but [not deprecated](http://hg.openjdk.java.net/jdk/jdk/file/e4fc1e578b3a/src/java.base/share/classes/java/util/Stack.java). Java still tries for backwards compatibility where possible so I wouldn't ever see them being removed given they've lasted the best part of 22 years not even being deprecated.

Also if you're using it for the thread safety you'll have to be careful the collection you replace it with is safe too (which `Deque` as an interface doesn't guarantee for example).

Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/providers/MinecraftAssetsProvider.java]: moved to a `ConcurrentLinkedDeque`, will test once im at home.

**Comment by modmuss50** [comment]
Tested again, im happy with the changes.

**Pull request merged**